### PR TITLE
[Fix] Remove policy from paginated pool candidates query

### LINKED ISSUE #10932: ♻️ Perf - `ScreeningAndEvaluation_Candidates` is slow
## ♻️ Debt/Refactor

The Screening And Evaluation Page queries often show up on the slow queries logs.  Let's improve the performance of the `ScreeningAndEvaluation_Candidates` query.

image

For now, lets do so by skipping the Policy check per user, and relying on the authorizedToView scope

## ✅ Acceptance Criteria

- [ ] Performance of ScreeningAndEvaluation_Candidates improved
- [ ] Skip policy check for every user
- [ ] Add some phpunit tests to ensure correct roles can see correct users
- [ ] Spin off ticket for improving pagination on kanban page

### PR DESCRIPTION
**Pull request by esizer**
🤖 Resolves #10932 

## 👋 Introduction

Removes the resolved policy on the pool candidates paginated query to improve performance.

## 🕵️ Details

This relies only on the authorized scope for a pool candidate when making a paginated query. Due to this, we had to make the policy more strict by checking to see if a user has any of the required permissions and if they do not, we simply exit early, returning an empty result.

## 🧪 Testing

1. Build `pnpm run dev:fresh`
2. Login with different roles
3. Navigate to the pool candidates table
4. Confirm you only see the expected candidates

### CODE REVIEW

File: api/app/Models/PoolCandidate.php
Comment (mnigh) [inline api/app/Models/PoolCandidate.php:1]: Not part of this PR, but noticed Line 1099 has a typo `UNSUFFCESSFUL`.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/5242204b48dc916e42e37f64caa07dca13de3b19/api/app/Models/PoolCandidate.php#L1080

File: api/app/Models/PoolCandidate.php [hunk 1]
```diff
@@ -589,10 +589,10 @@ public function scopeNotDraft(Builder $query): Builder
     public function getPoolCandidateStatusAttribute($candidateStatus)
     {
         // pull info
```
Comment (mnigh) [inline api/app/Models/PoolCandidate.php:591]: any reason to keep this and other commented out code?

Comment (esizer) [reply api/app/Models/PoolCandidate.php:591]: I actually have no clue :sweat_smile: It was causing issues with my linter so I commented it out since it wasn't being used but I wasn't sure if it was planned on being brought back or not :woman_shrugging:

File: api/app/Models/PoolCandidate.php
Comment (esizer) [reply api/app/Models/PoolCandidate.php:1]: easy enough change so I went ahead and made it [fix typo](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11022/commits/7ff9032558b778954f178cfc361402bcba908e87)

File: api/tests/Feature/PoolCandidatesPaginatedTest.php [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        $res->assertJsonFragment([
+            'paginatorInfo' => [
+                'total' => $count,
+            ],
+        ]);
+
+        foreach ($ids as $id) {
+            $res->assertJsonFragment(['id' => $id]);
+        }
+    }
```
Comment (yonikid15) [inline api/tests/Feature/PoolCandidatesPaginatedTest.php:198]: I know this works, but can we move this above the tests before `setUp()` since it's being used before.

File: api/app/Models/PoolCandidate.php (hunk 1 above)
Comment (esizer) [reply api/app/Models/PoolCandidate.php:591]: Added ticket to remove:

 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/11032

**Review by yonikid15 — APPROVED** [review]
Looks good just left a comment on the position of the query function in the test file

File: api/tests/Feature/PoolCandidatesPaginatedTest.php (hunk 2 above)
Comment (esizer) [reply api/tests/Feature/PoolCandidatesPaginatedTest.php]: Yup, moved up in [move assertion helper above first usage](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11022/commits/938e937881b382ae888fb141936d6909930a74b0)

File: api/tests/Feature/PoolCandidatesPaginatedTest.php [hunk 3]
```diff
@@ -34,7 +34,9 @@ class PoolCandidatesPaginatedTest extends TestCase
 
     public PoolCandidate $applicantCandidate;
 
-    public string $query = /** GraphQL */ '
+    public string $query =
+    /** GraphQL */
+    '
```
Comment (yonikid15) [inline api/tests/Feature/PoolCandidatesPaginatedTest.php:39]: Everything looks good! Merge after fixing the linting error. This indentation is causing the linting error btw.

Comment (esizer) [reply api/tests/Feature/PoolCandidatesPaginatedTest.php]: fixed in [fix lint error](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11022/commits/3eff274b92b79fa05eb952ce106bd8d17a1beabc)

**Comment by vd1992** [comment]
Given the issue has to deal with the performance of `ScreeningAndEvaluation_Candidates` in any fashion, I noticed another possible tweak while working on something else.
The query is fetching the pool relation for every single candidate. In an operation of say 500 candidates, that 500 instances of the same pool being fetched just for the sake of it. I am assuming for the sake of it since it is only getting `id`

image

image

**Comment by esizer** [comment]
> Given the issue has to deal with the performance of `ScreeningAndEvaluation_Candidates` in any fashion, I noticed another possible tweak while working on something else.
> The query is fetching the pool relation for every single candidate. In an operation of say 500 candidates, that 500 instances of the same pool being fetched just for the sake of it. I am assuming for the sake of it since it is only getting `id`

Good call, removed in [remove unecessary data fetch](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11022/commits/97d771d8703fcfcb38e046605a884f54e523f6c8)

**Comment by vd1992** [comment]
> > Given the issue has to deal with the performance of `ScreeningAndEvaluation_Candidates` in any fashion, I noticed another possible tweak while working on something else.
> > The query is fetching the pool relation for every single candidate. In an operation of say 500 candidates, that 500 instances of the same pool being fetched just for the sake of it. I am assuming for the sake of it since it is only getting `id`
> 
> Good call, removed in [remove unecessary data fetch](https://github.com/GCTC-NTGC/gc-digital-talent/pull/11022/commits/97d771d8703fcfcb38e046605a884f54e523f6c8)

That darn `PoolCandidate` type 😆

**Comment by yonikid15** [comment]
Just in case you are waiting on me to merge this,  took another look and it looks good. I created a new issue for an unrelated [bug](https://github.com/GCTC-NTGC/gc-digital-talent/issues/11035)

**Review by yonikid15 — APPROVED** [review]
